File: simulator/board.js
// ----- 領域関連 ----- start
// 領域クラス
// 現在値(初期値)と最大値を格納
class Area{
    constructor(val, max){
        this.val = val;
        this.max = max;
    }
}
// 変数定義
const distance = new Area(10, 10);
const dust = new Area(0, 100);
const life_0 = new Area(10, 100);
const aura_0 = new Area(3, 5);
const flare_0 = new Area(0, 100);
const life_1 = new Area(10, 100);
const aura_1 = new Area(3, 5);
const flare_1 = new Area(0, 100);
// 値を変化させる関数
// 返値：成功なら変更後の値，失敗なら-1
function chgAreaVal(area, n){
    area.val += n;
    if(area.val < 0 || area.val > area.max){
        area.val -= n;
        return -1;
    }
    return area.val;
}
// 結晶をn個移動させる関数 A→n→B
// 返値：移動個数
function moveAreaVal(areaA, areaB, n){
    if(chgAreaVal(areaA, -n) == -1){ //失敗した場合
        return 0;
    }
    if(chgAreaVal(areaB, n) == -1){ //失敗した場合
        chgAreaVal(areaA, n);
        return 0;
    }
    return n;
}
// 結晶を「できる限り」n個まで移動させる関数 A→n→B
// 返値：移動個数
function moveAreaValPoss(areaA, areaB, n){
    for(let i = 0; i < n; ++i){ //n回1つづつ移動
        if(moveAreaVal(areaA, areaB, 1) == 0){
            return i;
        }
    }
    return n;
}
// 表示関数
function outputBoard(){
    const output = 
            "P0 ライフ:" + life_0.val +
            "\nP0 オーラ:" + aura_0.val +
            "\nP0 フレア:" + flare_0.val +
            "\nP1 ライフ:" + life_1.val +
            "\nP1 オーラ:" + aura_1.val +
            "\nP1 フレア:" + flare_1.val +
            "\n間合 　　:" + distance.val +
            "\nダスト 　:" + dust.val;
    console.log(output);
}
// ----- 領域関連 ----- end

// ----- 監視 -----
// TODO:あとで移動させる
// オブジェクトのプロパティの値の変更を監視
// オブジェクト名, プロパティ名(文字列), 実行する関数
function watchValue(val, propName, func) {
    let value = val[propName]; // valueと監視対象を同一化
    Object.defineProperty(val, propName, {
        get: () => value, // 呼出時実行 valueを返す
        set: newValue => { // 値が変わる時の処理
            const oldValue = value;
            value = newValue;
            func(oldValue, newValue); // 旧値と新値が代入
        },
        configurable: true
    });
}
// プレイヤーのある領域を監視
function watchPlayers(players, area, func){
    watchValue(players[area], "val", func);
}
// ライフが0かの判定
function decisionLose(val0, val1){
    console.log(val0 + ", " + val1);
    if(val1 <= 0){ console.log("Player" +  + " Lose."); }
}

File: python/board.py
import MyPrint as myp

# 領域関連
# 領域クラス
# 現在値(初期値)と最大値を格納
class Area:
    def __init__(self, name, val, max) -> None:
        self.name = name # 領域名
        self.val = val
        self.max = max

# 値を変化させる関数
# 返値：成功なら変更後の値，失敗なら-1
def chgAreaVal(area, n):
    area.val += n
    if(area.val < 0 or area.val > area.max):
        area.val -= n
        return -1
    return area.val

# 結晶をn個移動させる関数 A→n→B
# 返値：移動個数, 失敗時-1
def moveAreaVal(areaA, areaB, n, log=True):
    areaAprev = areaA.val
    areaBprev = areaB.val
    if(chgAreaVal(areaA, -n) == -1): # 失敗した場合
        return -1
    if(chgAreaVal(areaB, n) == -1): # 失敗した場合
        chgAreaVal(areaA, n)
        return -1
    if log == True: # 移動ログ
        myp.printMove(f"{areaA.name}({areaAprev}→{areaA.val}) →{n}→ {areaB.name}({areaBprev}→{areaB.val})")
    return n

# 結晶を「できる限り」n個まで移動させる関数 A→n→B
# 返値：移動個数
def moveAreaValPoss(areaA, areaB, n):
    areaAprev = areaA.val
    areaBprev = areaB.val
    for i in range(n): # n回1つづつ移動
        if(moveAreaVal(areaA, areaB, 1, log=False) == -1): # 移動不可なら
            myp.printMove(f"{areaA.name}({areaAprev}→{areaA.val}) →{i}→ {areaB.name}({areaBprev}→{areaB.val})")
            return i
    # 全部移動したら↓
    myp.printMove(f"{areaA.name}({areaAprev}→{areaA.val}) →{n}→ {areaB.name}({areaBprev}→{areaB.val})")
    return n

# 表示関数
# 引数: [P0ライフ, P0オーラ, P0フレア, P1ライフ, P1オーラ, P1フレア, 間合, ダスト]
def outputBoard(areas):
    output = f"""P0 ライフ:{areas.life_0.val}
P0 オーラ:{areas.aura_0.val}
P0 フレア:{areas.flare_0.val}
P1 ライフ:{areas.life_1.val}
P1 オーラ:{areas.aura_1.val}
P1 フレア:{areas.flare_1.val}
間合 　　:{areas.distance.val}
ダスト 　:{areas.dust.val}"""
    print(output)

class Areas:
    def __init__(self) -> None:
        self.distance = Area("間合", 10, 10)
        self.dust = Area("ダスト", 0, 100)
        self.life_0 = Area("ライフP0", 10, 100)
        self.aura_0 = Area("オーラP0", 3, 5)
        self.flare_0 = Area("フレアP0", 0, 100)
        self.life_1 = Area("ライフP1", 10, 100)
        self.aura_1 = Area("オーラP1", 3, 5)
        self.flare_1 = Area("フレアP1", 0, 100)

# テスト
if __name__ == "__main__":
    # 変数定義例
    areas = Areas()

    outputBoard(areas)
    print(moveAreaValPoss(areas.distance, areas.dust, 8))
    outputBoard(areas)